colcon_core/verb/__init__.py

```diff
@@ -85,3 +85,33 @@ def check_and_mark_build_tool(build_base, *, this_build_tool='colcon'):
         os.makedirs(build_base, exist_ok=True)
 
     marker_path.write_text(this_build_tool + '\n')
+
+
+def check_and_mark_install_layout(install_base, *, merge_install):
+    """
+    Check the marker file for the previous install layout, otherwise create it.
+
+    The marker filename is `.colcon_install_layout`.
+
+    :param str install_base: The install directory
+    :param bool merge_install: The flag if all packages share the same prefix
+    :raises RuntimeError: if the marker file contains a different install
+      layout
+    """
+    this_install_layout = 'merged' if merge_install else 'isolated'
+    marker_path = Path(install_base) / '.colcon_install_layout'
+    if marker_path.parent.is_dir():
+        if marker_path.is_file():
+            previous_install_layout = marker_path.read_text().rstrip()
+            if previous_install_layout == this_install_layout:
+                return
+            change_option = 'remove' if merge_install else 'add'
+            raise RuntimeError(
+                "The install directory '{install_base}' was created with the "
+                "layout '{previous_install_layout}'. Please remove the "
+                'install directory, pick a different one or {change_option} '
+                "the '--merge-install' option.".format_map(locals()))
+    else:
+        os.makedirs(install_base, exist_ok=True)
```

Comment:
Could use a nicer error message if `install_base` is not a directory

```
$ touch not-a-dir
$ colcon build --install-base not-a-dir 
[0.458s] ERROR:colcon:colcon build: [Errno 17] File exists: 'not-a-dir'
Traceback (most recent call last):
  File "/home/sloretz/projects/colcon/install-merge/lib/python3.5/site-packages/colcon_core/command.py", line 371, in verb_main
    rc = context.args.main(context=context)
  File "/home/sloretz/projects/colcon/install-merge/lib/python3.5/site-packages/colcon_core/verb/build.py", line 170, in main
    merge_install=context.args.merge_install)
  File "/home/sloretz/projects/colcon/install-merge/lib/python3.5/site-packages/colcon_core/verb/__init__.py", line 115, in check_and_mark_install_layout
    os.makedirs(install_base, exist_ok=True)
  File "/usr/lib/python3.5/os.py", line 241, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: 'not-a-dir'
```

Reply:
Good point. I handled this in a2a6753d6eedb101baa7df9f02a15e47cd64330d as well as covered it by a test.